File: Tarefa2/src/Shader.cpp
#include "Shader.h"
#include <GL/glew.h>
#include <glm/gtc/type_ptr.hpp>
#include <fstream>
#include <iostream>
#include <sstream>
#include <cstdlib>

static std::string ReadFile (const std::string& filename) {
  std::ifstream fp(filename);
  if (!fp.is_open()) { std::cerr << "Could not open file: " << filename << std::endl; std::exit(1); }
  std::stringstream ss; ss << fp.rdbuf(); return ss.str();
}

static void CompileShader (const std::string& filename, GLuint id) {
  glCompileShader(id);
  GLint status; glGetShaderiv(id, GL_COMPILE_STATUS, &status);
  if (!status) {
    GLint len; glGetShaderiv(id, GL_INFO_LOG_LENGTH, &len);
    std::string log(len, '\0');
    glGetShaderInfoLog(id, len, 0, log.data());
    std::cerr << filename << ":\n" << log << std::endl;
    std::exit(1);
  }
}

static GLuint CreateShader (GLenum type, const std::string& filename) {
  GLuint id = glCreateShader(type);
  if (!id) { std::cerr << "Could not create shader object"; std::exit(1); }
  std::string src = ReadFile(filename); const char* csrc = src.c_str();
  glShaderSource(id,1,&csrc,nullptr); CompileShader(filename,id); return id;
}

ShaderPtr Shader::Make () { return ShaderPtr(new Shader()); }

Shader::Shader () { m_pid = glCreateProgram(); if (!m_pid) { std::cerr << "Could not create shader object"; std::exit(1);} }
Shader::~Shader () {}

void Shader::AttachVertexShader (const std::string& f) { glAttachShader(m_pid, CreateShader(GL_VERTEX_SHADER,f)); }
void Shader::AttachFragmentShader (const std::string& f) { glAttachShader(m_pid, CreateShader(GL_FRAGMENT_SHADER,f)); }
void Shader::AttachGeometryShader (const std::string& f) { glAttachShader(m_pid, CreateShader(GL_GEOMETRY_SHADER,f)); }
void Shader::AttachTesselationShader (const std::string& c, const std::string& e) {
  glAttachShader(m_pid, CreateShader(GL_TESS_CONTROL_SHADER,c));
  glAttachShader(m_pid, CreateShader(GL_TESS_EVALUATION_SHADER,e));
}

void Shader::Link () {
  glLinkProgram(m_pid);
  GLint status; glGetProgramiv(m_pid, GL_LINK_STATUS,&status);
  if (!status) {
    GLint len; glGetProgramiv(m_pid, GL_INFO_LOG_LENGTH,&len);
    std::string log(len,'\0'); glGetProgramInfoLog(m_pid,len,0,log.data());
    std::cerr << log << std::endl; std::exit(1);
  }
}

void Shader::UseProgram () const { glUseProgram(m_pid); }

void Shader::SetUniform (const std::string& n, int x) const { GLint loc = glGetUniformLocation(m_pid,n.c_str()); glUniform1i(loc,x); }
void Shader::SetUniform (const std::string& n, float x) const { GLint loc = glGetUniformLocation(m_pid,n.c_str()); glUniform1f(loc,x); }
void Shader::SetUniform (const std::string& n, const glm::vec3& v) const { GLint loc = glGetUniformLocation(m_pid,n.c_str()); glUniform3fv(loc,1,glm::value_ptr(v)); }
void Shader::SetUniform (const std::string& n, const glm::vec4& v) const { GLint loc = glGetUniformLocation(m_pid,n.c_str()); glUniform4fv(loc,1,glm::value_ptr(v)); }
void Shader::SetUniform (const std::string& n, const glm::mat4& m) const { GLint loc = glGetUniformLocation(m_pid,n.c_str()); glUniformMatrix4fv(loc,1,GL_FALSE,glm::value_ptr(m)); }
void Shader::SetUniform (const std::string& n, const std::vector<int>& v) const { GLint loc = glGetUniformLocation(m_pid,n.c_str()); glUniform1iv(loc,v.size(),v.data()); }
void Shader::SetUniform (const std::string& n, const std::vector<float>& v) const { GLint loc = glGetUniformLocation(m_pid,n.c_str()); glUniform1fv(loc,v.size(),v.data()); }
void Shader::SetUniform (const std::string& n, const std::vector<glm::vec3>& v) const { GLint loc = glGetUniformLocation(m_pid,n.c_str()); glUniform3fv(loc,v.size(),(float*)v.data()); }
void Shader::SetUniform (const std::string& n, const std::vector<glm::vec4>& v) const { GLint loc = glGetUniformLocation(m_pid,n.c_str()); glUniform4fv(loc,v.size(),(float*)v.data()); }
void Shader::SetUniform (const std::string& n, const std::vector<glm::mat4>& v) const { GLint loc = glGetUniformLocation(m_pid,n.c_str()); glUniformMatrix4fv(loc,v.size(),GL_FALSE,(float*)v.data()); }

File: Tarefa3/scene_graph/python/shader.py
from OpenGL.GL import *
import numpy as np
import glm

import shaderutl as sutl

# read file to a string
class Shader:
  def __init__ (self, light=None, space="camera"):
    self.shaders = []
    self.textunit = 0
    self.light = light
    self.space = space
    self.pid = None

  def AttachVertexShader (self, filename):
    self.shaders.append(sutl.create_shader(GL_VERTEX_SHADER,filename))

  def AttachFragmentShader (self, filename):
    self.shaders.append(sutl.create_shader(GL_FRAGMENT_SHADER,filename))

  def AttachGeometryShader (self, filename):
    self.shaders.append(sutl.create_shader(GL_GEOMETRY_SHADER,filename))

  def AttachTesselationShader (self, control_filename, evaluation_filename):
    self.shaders.append(sutl.create_shader(GL_TESS_CONTROL_SHADER,control_filename))
    self.shaders.append(sutl.create_shader(GL_TESS_EVALUATION_SHADER,evaluation_filename))
  
  def Link (self):
    self.pid = sutl.create_program(*self.shaders)

  def GetLight (self):
    return self.light

  def GetLightingSpace (self):
    return self.space

  def UseProgram (self):
    type(self.pid)
    glUseProgram(self.pid)

  def SetUniform (self, varname, x):
    loc = glGetUniformLocation(self.pid,varname)
    tp = type(x)
    if tp == int:
      glUniform1i(loc,x)
    elif tp == float:
      glUniform1f(loc,x)
    elif tp == glm.vec3:
      glUniform3fv(loc,1,glm.value_ptr(x))
    elif tp == glm.vec4:
      glUniform4fv(loc,1,glm.value_ptr(x))
    elif tp == glm.mat4x4:
      glUniformMatrix4fv(loc,1,GL_FALSE,glm.value_ptr(x))
    elif tp == list:
      tpe = type(x[0])
      if tpe == int:
        glUniform1iv(loc,len(x),np.array(x,dtype='int'))
      elif tpe == float:
        glUniform1fv(loc,len(x),np.array(x,dtype='float'))
      elif tpe == glm.vec3:
        glUniform3fv(loc,len(x),np.array(x,dtype='float'))
      elif tpe == glm.vec4:
        glUniform4fv(loc,len(x),np.array(x,dtype='float'))
      elif tpe == glm.mat4x4:
        glUniformMatrix4fv(loc,len(x),GL_FALSE,np.array(x,dtype='float'))
      else:
        raise SystemError("Type not supported in list in Shader.SetUniform: " + str(tpe))
    else:
      raise SystemError("Type not supported in Shader.SetUniform: " + str(tp))
    
  def ActiveTexture (self, varname):
    self.SetUniform(varname,self.texunit)
    glActiveTexture(GL_TEXTURE0+self.texunit)
    self.texunit += 1
  
  def DeactiveTexture (self):
    self.texunit -= 1

  def GetCurrentUnit (self):
    return self.texunit

  def Load (self, st):
    st.PushShader(self)
    if (self.light):
      self.light.Load(st)

  def Unload (self, st):
    st.PopShader()

